src/core/net/dns_types.cpp

```diff
@@ -221,20 +222,38 @@ Error Name::AppendTerminator(Message &aMessage)
 
 Error Name::AppendPointerLabel(uint16_t aOffset, Message &aMessage)
 {
+    Error    error;
+    uint16_t value;
+
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE && !OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE
+    if (Instance::Get().IsUncompressedDnsNameEnabled())
+    {
+        // If "uncompressed DNS name" mode is enabled, instead of
+        // appending the pointer label, read the name from the message
+        // and append it uncompressed. Note that the `aOffset` parameter
+        // in this method is given relative to the start of DNS header
+        // in `aMessage` (which `aMessage.GetOffset()` specifies).
+
+        error = Name(aMessage, aOffset + aMessage.GetOffset()).AppendTo(aMessage);
+        ExitNow();
+    }
+#endif
+
     // A pointer label takes the form of a two byte sequence as a
     // `uint16_t` value. The first two bits are ones. This allows a
     // pointer to be distinguished from a text label, since the text
     // label must begin with two zero bits (note that labels are
     // restricted to 63 octets or less). The next 14-bits specify
     // an offset value relative to start of DNS header.
 
-    uint16_t value;
-
     OT_ASSERT(aOffset < kPointerLabelTypeUint16);
 
     value = HostSwap16(aOffset | kPointerLabelTypeUint16);
 
-    return aMessage.Append(value);
+    ExitNow(error = aMessage.Append(value));
```

Comment:
From code flow, no it is not necessary. Added it so that when `REFERENCE_DEVICE` config is not enabled, we don't get the error that the `exit` label is not used. 
- Alternative is to add `#if` around the `exit:` label (which looks kind of ugly).  
- Or another way is to not use `exit:` at all and have the whole block in an  `else` for the `if (Instance::Get().IsUncompressedDnsNameEnabled())`.

I felt adding `ExitNow()` at the end may be the cleanest way, but I don't have any strong preferences here. 

@jwhui, what do you think?

Reply:
Ah, yes, that makes sense. LGTM, thanks!